### ISSUE TITLE
TensorBoard scalars, images, distributions, histograms legend/description

### ISSUE THREAD

**Issue by piojanu** [google-research/planet#10]
Hi again!

I was able to run your code for adapted Sokoban environment (#9). Now I wonder what all of those statistics in TensorBoard mean. Would you be able to describe in one sentence the most important positions from scalars, images, distributions, histograms tabs? What am I looking at or maybe what SHOULD I look at?

Thanks for your time, I really appreciate that you (and whole team) open source your work and devote time to help the community work on their projects.

**Comment by piojanu** [comment]
Also, what does this statistic in the console mean: `step/score/loss/zs_entropy/zs_divergence =  [120200, -nan, 11374.6328, 37.9262047, 2.77373099]`. Is that `-nan` a problem?

**Comment by danijar** [comment]
Nice! The nan is not a problem at all, it just means that there is no score during iterations where no planning happens. The most important summaries at the "cem" scalar summaries that show the planning performance. If any of the "divergence" scalar summaries is at zero the `divergence_scale` is too high. Besides that, you can look at the "openloop" image summaries to see future frames imagined by the agent and at the "cem" image summaries to see the planning policy in the environment.

**Comment by piojanu** [comment]
I'll reread the paper (and maybe refresh knowledge on VAE) and look deeper into code to get back with better questions ;)
For now: (1) What does open/closedloop mean? (2) Also, I cannot find "cem" statistics beside return (e.g. there are no images with "cem" in their name). It learns something sensible, train images looks good, at least posterior ones. (3) Do I understand it correctly that prior images are just decoded noise (prior is normal distribution and it's now trained, right?).

**Comment by danijar** [comment]
Yes, I suggest looking at the paper again. The closed loop prior is a one-step prediction and the closed loop posterior is a 0-step reconstruction. Open loop means predicting forward for multiple steps without receiving intermediate image inputs. There should be a cem image summary once the agent has collected any test episodes.

**Comment by piojanu** [comment]
Thank! I'm discovering the paper once again, everything starts to make sense 👍 Also, here is the paper I was looking for:  [TensorFlow Agents: Efficient Batched Reinforcement Learning in TensorFlow](https://arxiv.org/pdf/1709.02878.pdf) 😄 Great job!

**Comment by piojanu** [comment]
Hi again! Do I understand it correctly that those two graphs:
![image](https://user-images.githubusercontent.com/23562932/55487966-676a9280-562f-11e9-9e75-816ea65e3aa6.png)
show train and test returns? Also, after your last updates in images there are those two summaries:
![image](https://user-images.githubusercontent.com/23562932/55488111-b1537880-562f-11e9-8088-76b0bee9212e.png)
but gif doesn't move and the second one "image" doesn't show anything :/ Should I open issue or I don't know something?

**Comment by danijar** [comment]
The GIF will only work if `ffmpeg` is available in your PATH. Otherwise, it falls back to showing the first frame as a static image. The second summary should show an image but it may be very wide (based on the sequence length) compared to its height. TensorBoard allows you to increase the image size which should make it visible.

**Comment by piojanu** [comment]
Okay, I've fixed ffmpeg (I'm able to do `ffmpeg -version` without error). We will see if now gif will work ;) But still, I wasn't able to show `graph/summaries/simulation/should_simulate_gym_sokoban/summary-gym_sokoban-cem-12/image/image/0`. Where do I need to click to increase the image size to make it visible?

**Comment by danijar** [comment]
![screenshot](https://user-images.githubusercontent.com/2111293/55581069-f936d600-56e9-11e9-9fa1-4c6016e067bb.png)

**Comment by piojanu** [comment]
@danijar do I understand it correctly, the test step log is mean score after `test_steps` runs of the game? Here is the definition of the test phase:
```
      trainer.add_phase(
          'test', config.test_steps, score, summary,
          batch_size=config.batch_shape[0],
          report_every=config.test_steps,
          log_every=config.test_steps,
          checkpoint_every=config.test_checkpoint_every)
```
I've looked through the code and I see that there is some running average kept, but I'd like to confirm that.

**Comment by piojanu** [comment]
I ask because test scores seems to have a lot of variance from global step to step.

**Comment by danijar** [comment]
It takes an average over the episodes simulation per test phase. Typically, there is only one simulation per test phase so, it's just a single score.

**Comment by piojanu** [comment]
How is that? Could you point to some line in code? E.g. here score is defined as mean od returns:
https://github.com/google-research/planet/blob/9cd9abed5b9a8831388f4d9da16e5604cfbd7c20/planet/training/define_summaries.py#L98-L113
And I thought it's run at every test step (so e.g. 100 times) and then in trainer those 100 runs are averaged. Could you highlight more how it works?

**Comment by danijar** [comment]
Sorry, it takes the mean over episodes per test step, not per test phase. So if the test phase spans multiple steps, there will be multiple scores on TensorBoard. The mean only changes anything if the episodes terminate early and multiple episodes fit within the `task.max_length` steps or if the simulation batch size is changed to something other than 1.

**Comment by piojanu** [comment]
I don't think so @danijar. I see in TensorBoard only one summary per test phase every. Could you double check on that? I got confused now, I can't see it in the code. E.g. what is for the `tools.StreamingMean` in `_define_step` method in `trainer.py` file lines from 256 to 288? It seems like running average of the score which gets logged to the TensorBoard every `log_every` steps which are passed as the parameter to the trainer and is set to `test_steps` in case of test phase (which would indicate that scores are logged ones every test phase).

**Comment by danijar** [comment]
The `report_every` parameter in the training loop tells at which interval summaries are computed. The `log_every` parameter tells at which frequency they are stored as summary. The scores are always averages since the last summary.

Since the code uses `report_every=config.test_steps` and `log_every=config.test_steps` the summaries are computed once per test phase (which means one episode is collected, at least assuming fixed episode lengths) and are also stored as summary once per test phase.

Thus, as mentioned earlier, the streaming mean does not change anything and the test scores are not averages. The streaming mean is only used to compute training summaries (e.g. loss values). You can find all of this specified in [utilities.py](https://github.com/google-research/planet/blob/9cd9abed5b9a8831388f4d9da16e5604cfbd7c20/planet/training/utility.py#L171).

**Comment by piojanu** [comment]
I see it, thanks. But please note that scores are calculated only `log_every` steps. This is `should_summarize` in `define_model.py`.
More specifically summaries are calculated only `log_every` steps (see it in line 141-144 in `define_model`). And score is also calculated in the same place `log_every` steps, otherwise it's set to 0. So this whole code for averaging scores doesn't make any sense for me. It doesn't work for case other than `log_every` = `report_every`.

Moreover, those training summaries in `utilities.py` are only printed. Those are not losses in TensorBoard. Summaries in TensorBoard are just values from one step (one per every train phase by default).

Could you confirm that? For me this logic is some legacy code that doesn't make sense anymore.

**Comment by danijar** [comment]
Your explanation sounds reasonable. I'll confirm if everything still works with `config.test_steps = 1`. Then it also makes more sense to set `log_every=1` and `report_every=1` for the test phase.

**Comment by piojanu** [comment]
@danijar thank you, I'm waiting for your results. To be honest, I should test it already, sorry for adding you more work. I have pretty limited resources, so it would be great if you could confirm this for me! 😄

**Comment by piojanu** [comment]
@danijar do you have an update on that one?

**Comment by danijar** [comment]
I found time to look into this. The test steps should be as in the code (100), or at least equal the batch size (50). This is because the step counter is advanced by the batch size after every processed data batch, so the step counts training examples (sequence chunks in this case). If you set test_steps to 1 it will only compute test summaries every 50 training phases.

**Comment by piojanu** [comment]
Thanks❣️
